pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,387 @@
+	AddHeaders(...*block.Header) error
+	HeaderHeight() uint32
+}
+
+// Service is a service that fetches blocks from NeoFS.
+type Service struct {
+	started           atomic.Bool
+	log               *zap.Logger
+	cfg               config.NeoFSBlockFetcher
+	stateRootInHeader bool
+
+	chain    Ledger
+	client   *client.Client
+	putBlock func(*block.Block) error
+	account  *wallet.Account
+
+	oidsCh   chan OidChannel
+	blocksCh chan *block.Block
+	wg       sync.WaitGroup
+
+	quit             chan struct{}
+	shutdownCallback func()
+}
+
+// New creates a new BlockFetcherService.
+func New(chain Ledger, cfg config.NeoFSBlockFetcher, logger *zap.Logger, putBlock func(*block.Block) error, shutdownCallback func()) (*Service, error) {
+	var (
+		account *wallet.Account
+		err     error
+	)
+
+	if cfg.UnlockWallet.Path != "" {
+		walletFromFile, err := wallet.NewWalletFromFile(cfg.UnlockWallet.Path)
+		if err != nil {
+			return nil, err
+		}
+		for _, acc := range walletFromFile.Accounts {
+			if err := acc.Decrypt(cfg.UnlockWallet.Password, walletFromFile.Scrypt); err == nil {
+				account = acc
+				break
+			}
+		}
+		if account == nil {
+			return nil, errors.New("failed to decrypt any account in the wallet")
+		}
+	} else {
+		account, err = wallet.NewAccount()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &Service{
+		chain: chain,
+		log:   logger,
+		quit:  make(chan struct{}),
+		cfg:   cfg,
+
+		putBlock:          putBlock,
+		account:           account,
+		stateRootInHeader: chain.GetConfig().StateRootInHeader,
+		shutdownCallback:  shutdownCallback,
+
+		// Use buffer of two batch sizes to load OIDs in advance:
+		//  * first full block of OIDs is processing by Downloader
+		//  * second full block of OIDs is available to be fetched by Downloader immediately
+		//  * third half-filled block of OIDs is being collected by OIDsFetcher.
+		oidsCh: make(chan OidChannel, 2*cfg.OIDBatchSize),
+
+		// Use buffer of a single OIDs batch size to provide smooth downloading and
+		// avoid pauses during blockqueue insertion.
+		blocksCh: make(chan *block.Block, cfg.OIDBatchSize),
+	}, nil
+}
+
+// Start runs the block fetcher service.
+func (bfs *Service) Start() error {
+	if !bfs.started.CompareAndSwap(false, true) {
+		return nil
+	}
+	bfs.log.Info("starting NeoFS block fetcher service")
+
+	var err error
+	bfs.client, err = neofs.GetSDKClient(context.Background(), bfs.cfg.Addresses[0])
+	if err != nil {
+		return err
+	}
+
+	// Start OIDs downloader routine.
+	bfs.wg.Add(1)
+	go bfs.oidDownloader()
+
+	// Start the set of blocks downloading routines.
+	for i := 0; i < bfs.cfg.DownloaderWorkersCount; i++ {
+		bfs.wg.Add(1)
+		go bfs.blockDownloader()
+	}
+
+	// Start routine that puts blocks into bQueue.
+	bfs.wg.Add(1)
+	go bfs.blockPutter()
+
+	return nil
+}
+
+// oidDownloader runs the appropriate blocks OID fetching method based on the configuration.
+func (bfs *Service) oidDownloader() {
+	defer bfs.wg.Done()
+	var err error
+
```

Comment:
Useless line. Maybe add it after `defer` because this `err` belongs to `if` below.